CoRIM store migration (#383)
veraison/services: scheme/README.md

Author: added.

only needs to be given callback validators for individual elements (such as
environments). E.g.

```go
import (
	"github.com/veraison/corim/comid"
    "github.com/veraison/services/scheme/common"
)

func revValEnvValidator(e *comid.Environment) error {
    // validate reference value environment
    return nil
}

func taEnvValidator(e *comid.Environment) error {
    // validate trust anchor environment
    return nil
}

func measurementsValidator(meas []comid.Measurement) error {
    // validate reference value measurements
    return nil
}

func cryptoKeysValidator(keys []*comid.CryptoKeys) error {
    // validate attestation verification keys
    return  nil
}

var validator = &common.TriplesValidator{
    TAEnviromentValidator: taEnvValidator,
    RefValEnviromentValidator: refValEnvValidator,
    // Alternatively to the above, EnvironmentValidator may specified, which
    // will be invoked for both trust anchors and reference values.
    CryptoKeysValidator: measurementsValidator,
    MeasurementsValidator: cryptoKeysValidator,
}

func init() {
	extMap := extensions.NewMap().Add(comid.ExtTriples, validator)

    // ...
}
```

> Yo do not need to specify all the validator functions listed above -- just the

Reviewer: ```suggestion
You do not need to specify all the validator functions listed above -- just the
```

In general, it is recommended that new schemes are defined using
`SchemeDescriptor` and `ISchemeImplementation` plus any additional methods from
`ISchemeHandler` that may be needed (such as `GetReferenceValueIDs`). 

It is typically not necessary to fully implement `ISchemeHandler` -- any of its
methods that are not needed for a particular implementation, and that are not
part of `ISchemeImplementation`, may be omitted. 

If, for whatever reason, this is not sufficient, it is possible to implement
`ISchemeHandler` directly:

```go
package main

import (
	"github.com/veraison/services/handler"
	"github.com/veraison/services/plugin"
)

type Handler struct {}

// ...
// Implementation of ISchemeHandler for Handler
// ...

func main() {
	handler.RegisterSchemeHandler(&Handler{})
	plugin.Serve()
}
```

(Note that `ISchemeHandler` embeds `ISchemeImplementation`, so you would still
need to implement it.)

### Call order

During evidence appraisal, `ISchemeHandler` methods will be invoked in the
following order:

1. `GetAttestationScheme`
2. `GetTrustAnchorIDs`
3. `ExtractClaims`
4. `GetReferenceValueIDs`
5. `ValidateEvidenceIntegrity`
6. `AppraiseClaims`

> If a particular `ISchemeImplementation` does not provided an

Reviewer: ```suggestion
If a particular `ISchemeImplementation` does not provide an
```

If the error occurs during a call to `ValidateEvidenceIntegrity` or sooner,
all claims inside the attestation result returned by VTS will be set to "crypto
validation failed".

If the error occurs during `AppraiseClaims`, the attestation result returned
from the call will be forwarded as-is. This means it is the responsibility of
`AppraiseClaims` to populate the claims within the attestation result it
returns appropriately, prior to returning the error.

Any error other than `BadEvidenceError` will result in all claims inside the
returned attestation result being set to "verifier malfunction", and the error
being propagated to the VTS client.

## Guidance on structuring and using CoRIMs in attestation schemes

> [!NOTE]
> This section should viewed as supplementing section 5.1 of
> [draft-ietf-rats-corim-09]. It describes how the CoRIMs are used by the
> Veraison Trusted Services (VTS) and therefore how they may be used by
> attestation schemes.

VTS deals with endorsements (reference measurements) and trust anchors as basic
units of provisioned data. This maps directly to `reference-triples` and
`attest-key-triples` described in section 5.1.4 of [draft-ietf-rats-corim-09].

> Triples relevant to a particular attestation are identified based on their

Author: This is talking about a particular attestation, i.e. one loop through the verification pipeline, not the attestation scheme in general; I'll replace "attestation" with "appraisal"

environments by matching them to the environments generated from the
attestation evidence during verification flow.

Keep in mind the following things when structuring CoRIMs for use with Veraison
Services:

- CoRIMs must contain _only_ CoMID tags; no other tags are currently supported.
  A CoRIM may contain any number of CoMIDs.
- Any information relevant to attestation verification must be contained
  _solely_ within `reference-triples` and `attest-key-triples`
  (`Triples.ReferenceValues` and `Triples.AttestVerifKeys` in the Go package)
  [^1]. A CoMID may contain any number of these triples.
- Meaning must NOT be assigned to the grouping of triples within a CoMID, or
  grouping of CoMIDs within a CoRIM. This grouping information is not available
  to the attestation schemes -- all relevant triples are retrieved as a single
  list.
- Meaning CAN be assigned to grouping of measurements within a single reference
  value triple, or to the grouping of keys within a single attest. verif. key
  triple.

> - Any information needed to match a triple to evidence must be contained

Reviewer: ```suggestion
- Any information needed to identify a triple from evidence must be contained
```

## Implementing Attestation Scheme Support

> [!NOTE]
> `example` sub-directory contains a "template" for new scheme boiler plate.
> You can start implementing a new scheme by copying it and filling in the
> TODO's.

An implementation of an attestation scheme needs to provide two things:

- A [`SchemeDescriptor`](../handler/schemedescriptor.go) providing basic
  information about the scheme.
- An implementation of
  [`ISchemeImplementation`](../handler/ischemeimplementation.go) interface.

Both of these need to then be registered

> - as a plugin by calling `handler.RegisterSchemeImpalementation`, and

Reviewer: ```suggestion
- as a plugin by calling `handler.RegisterSchemeImplementation`, and
```
